Modified horizontal layout to allow mirroring of the default layout, added new config value to toggle behavior on/off

smithbm2316 opened the pull request:
This is a feature that I have been wanting for a while now in Telescope, but never really got around to opening an issue for it for whatever reason. Finally decided I might as well see if I can figure out how to do it myself by poking around the codebase 🙂. I would love if the horizontal layout allowed for a config option to let you mirror the default horizontal layout. Here's a visual:

Current horizontal layout (new config option would be set to `false`):
```
   +-----------------+---------------------+
   |                 |                     |
   |     Results     |                     |
   |                 |       Preview       |
   |                 |                     |
   +-----------------|                     |
   |     Prompt      |                     |
   +-----------------+---------------------+
```
Horizontal layout with the new config option set to `true`:
```
   +-------------------+-------------------+
   |                   |                   |
   |                   |       Results     |
   |      Preview      |                   |
   |                   |                   |
   +                   |------------------ |
   |                   |       Prompt      |
   +-------------------+-------------------+
```

I am by no means tied to the name that I chose for this config option, nor tied to the way that I implemented this. I just would love if this was a possibility in telescope, as I'd prefer to see the `prompt` and `results` windows on the right side instead of the left. Thanks for all your hard work on Telescope, it's such an invaluable tool to my current workflow!

elianiva commented:
LGTM. Maybe you could also make the mirrored version of the vertical layout, it will solve #438 minus refactoring.

smithbm2316 commented:
> LGTM. Maybe you could also make the mirrored version of the vertical layout, it will solve #438 minus refactoring.

I could certainly look into doing that! I'd imagine that it would be good to have the vertical and horizontal mirroring done separately, so that if someone is using the `flex` layout, they can choose whether they want the mirroring to affect only when the flex layout is in horizontal, vertical, or both modes. Maybe a new value like `mirror_layout = { "horizontal" = false, "vertical" = false }` as the defaults? I'll look into adding vertical mirroring to solve #438 in this PR as well.

elianiva commented:
Yes, separate config flag for horizontal or vertical would be good.

Conni2461 commented on lua/telescope/config.lua:
```diff
@@ -37,6 +37,7 @@ function config.set_defaults(defaults)
   set("scroll_strategy", "cycle")
 
   set("layout_strategy", "horizontal")
+  set("layout_mirror_horizontal", false)
```
We could have this part of layout_defaults:
```lua
layout_defaults = {
  horizontal = {
    width_padding = 0.1,
    height_padding = 0.1,
    preview_width = 0.6,
    mirror = false,
  },
  vertical = {
    width_padding = 0.05,
    height_padding = 1,
    preview_height = 0.5,
    mirror = true
  }
},

```

Conni2461 replied:
And probably should

smithbm2316 replied:
That makes sense, I had figured that might be why you had the `layout_defaults` option as a TODO when I was deciding what to name the config option. Do you want me to implement allowing a config for `width_padding`, `height_padding`, and `preview_height` as well as well as adding the mirroring for the vertical layout?

Conni2461 replied:
Oh wait they are just documented with TODO? Thats seems wrong. So they already exists. I just proposed to add mirror to that table.
You just need to add it to this table and then it should be available there.
https://github.com/nvim-telescope/telescope.nvim/blob/315e73698067ad96f71f1a13ee8b6c7567e61c34/lua/telescope/pickers/layout_strategies.lua#L64-L68

Like here `layout_config.mirror or default`
https://github.com/nvim-telescope/telescope.nvim/blob/315e73698067ad96f71f1a13ee8b6c7567e61c34/lua/telescope/pickers/layout_strategies.lua#L77

smithbm2316 replied:
![image](https://user-images.githubusercontent.com/37027883/109066689-54f71780-76e5-11eb-8e76-5f336c17a4f5.png)
Apologies, I misspoke. The `layout_defaults` option is marked as TODO in the README for telescope, not actually in the codebase itself (at least that I could find)

smithbm2316 replied:
@Conni2461 when you get a chance, take a look at the latest commit. I think I have fixed the confusion. I added `mirror` to the `layout_config` table and added a default `mirror` option of `false` to the `layout_defaults` table, for both `horizontal` and `vertical` layouts.

smithbm2316 commented:
Alright, I'll have to clean up this git history a bit later, because I ran into a number of luacheck warnings. I should probably figure out how to get luacheck integrated into my neovim editor as well too so I don't find out about these errors until I commit. 

Has anyone seen this error before?
```
Testing: 	/home/runner/work/telescope.nvim/telescope.nvim/lua/tests/automated/pickers/scrolling_spec.lua	
Fail	||	scrolling strategies should handle cycling for full list	
            .../site/pack/vendor/start/plenary.nvim/lua/plenary/job.lua:465: 'nvim --noplugin -u scripts/minimal_init.vim -c lua require("telescope.pickers._test")._execute("./lua/tests/pickers/find_files__scrolling_descending_cycle.lua")' was unable to complete in 10000 ms
            
            stack traceback:
            	./lua/telescope/pickers/_test.lua:158: in function 'get_results_from_file'
            	./lua/telescope/pickers/_test.lua:217: in function 'run_file'
            	...cope.nvim/lua/tests/automated/pickers/scrolling_spec.lua:7: in function <...cope.nvim/lua/tests/automated/pickers/scrolling_spec.lua:6>
```
I don't really have much of a clue what the error it's throwing is. I found a previous PR #387 had a similar issue that was unresolved there. It apparently might be a plenary bug? I'm not sure.

tjdevries commented:
Don't worry about cleaning up git history, we'll just squash when we merge.

tjdevries commented on lua/telescope/pickers/layout_strategies.lua:
```diff
@@ -261,9 +274,15 @@ layout_strategies.vertical = function(self, max_columns, max_lines)
-    preview.line = height_padding
-    results.line = preview.line + preview.height + 2
-    prompt.line = results.line + results.height + 2
+    if not config.values.layout_defaults.vertical.mirror then
```
Shouldn't this use `layout_config` not `config.values`?

smithbm2316 replied:
When I tried `layout_config` instead of `config.values`, it was returning a `nil` value to me. I can double check that I didn't make a mistake there, but it wasn't working until I switched to using `config.values`.

smithbm2316 commented:
Thanks everyone for the feedback, I think it should all be working properly now. Apologies for the original issues, I had originally just been playing around with the copy of Telescope installed by `packer.nvim`, since I had never manually installed a plugin before,  didn't think it would be so easy to implement, and didn't know you could point `packer.nvim` to a manually installed plugin on your system outside of neovim's runtime path. So when I went to fork the repo, I didn't end up pushing the exactly correct changes to this PR, and caused some extra confusion. I think everything should be in order now and should also solve #438 as well! Thank you all for your patience and for being so responsive as well, you're a very welcoming community to someone as new to Lua development as myself 😃.

elianiva commented on lua/telescope/pickers/layout_strategies.lua:
```diff
@@ -218,7 +226,12 @@ end
+  local layout_config = validate_layout_config(self.layout_config or {}, {
+    width_padding = "How many cells to pad the width",
+    height_padding = "How many cells to pad the height",
+    preview_width = "(Resolvable): Determine preview width",
```
Shouldn't this be `preview_height` because it's vertical?

smithbm2316 replied:
Yes, that is fixed now

tjdevries commented on lua/telescope/config.lua:
```diff
@@ -37,7 +37,15 @@ function config.set_defaults(defaults)
 
   set("layout_strategy", "horizontal")
-  set("layout_defaults", {})
+  local layout_defaults_table = {
+    horizontal = {
+      mirror = false,
+    },
+    vertical = {
+      mirror = false,
+    },
+  }
+  set("layout_defaults", { layout_defaults_table })
```
I don't think this really does anything -- since the defaults are false, they'll come out `falsey` when we check them later. I don't know that you need to do this.

smithbm2316 replied:
Oh interesting, I didn't even think about that. It does look like removing that change altogether still makes my updates work properly!

Conni2461 reviewed (approved):
I just have to go to bed and things just happen how i hoped they :)

LGTM thanks and works like a charm i gonna mirror vertical (@tami5 this implements (prompt -> results -> preview for vertical).

We might want to add `mirror` key to the documentation in this file. TJ just added them with his docgen PR. You have to rebase and then add it to the brief thanks :D

smithbm2316 commented:
> I just have to go to bed and things just happen how i hoped they :)
> 
> LGTM thanks and works like a charm i gonna mirror vertical (@tami5 this implements (prompt -> results -> preview for vertical).
> 
> We might want to add `mirror` key to the documentation in this file. TJ just added them with his docgen PR. You have to rebase and then add it to the brief thanks :D

I have another project for college that I'll need to be focusing most of my efforts towards today and tomorrow, but hopefully this weekend I can look into adding some docs for this :smiley:

smithbm2316 commented:
Hey, added some initial docs here. I can probably add some more to make it a bit more clear, but I'm not sure what the default values for `width_padding`, `height_padding`, and `preview_width` are, so I left those unchanged in the README and in the docs. I also updated the [wiki's Layout page](https://github.com/nvim-telescope/telescope.nvim/wiki/Layouts#layout-defaults) to reflect the new changes here, and linked to that wiki page in the README. Will probably add some more docs tomorrow as well.

Conni2461 commented:
Oh thats really good. :) thanks. Currently you still need to do some manual wrapping and indenting to make the output look good. So i wait until you address these issues :) You can see the generated output here: https://github.com/nvim-telescope/telescope.nvim/pull/548/commits/1186cd8ab88c4877c45fa0a9a98e6a8690d07f16

Also lualint is not happy: https://github.com/nvim-telescope/telescope.nvim/runs/1997599350

After that is sorted out i am happy to press the button :D

smithbm2316 commented:
> Oh thats really good. :) thanks. Currently you still need to do some manual wrapping and indenting to make the output look good. So i wait until you address these issues :) You can see the generated output here: [1186cd8](https://github.com/nvim-telescope/telescope.nvim/commit/1186cd8ab88c4877c45fa0a9a98e6a8690d07f16)
> 
> Also lualint is not happy: https://github.com/nvim-telescope/telescope.nvim/runs/1997599350
> 
> After that is sorted out i am happy to press the button :D

I...really need to get a whitespace trimmer set up in my Neovim :laughing:. I think the lualint warnings should be sorted now, forgot to run that locally on my commit before pushing, apologies.

Conni2461 commented:
I am talking about the generated telescope.txt.

It should be wrapped at 80 line length. but that is currently not working. So i am not sure how to handle that now because tj also pushed a `TODO:` with a line length of 100.
I think i should just send a pic :rofl: 

Currently:
![image](https://user-images.githubusercontent.com/15233006/109414728-b0125e00-79b4-11eb-9223-0b3f141fa32b.png)
And that is how i think it should look like:
![image](https://user-images.githubusercontent.com/15233006/109414743-d0dab380-79b4-11eb-9884-8775fc47f1e3.png)

smithbm2316 commented:
> I am talking about the generated telescope.txt.
> 
> It should be wrapped at 80 line length. but that is currently not working. So i am not sure how to handle that now because tj also pushed a `TODO:` with a line length of 100.
> I think i should just send a pic rofl
> 
> Currently:
> ![image](https://user-images.githubusercontent.com/15233006/109414728-b0125e00-79b4-11eb-9223-0b3f141fa32b.png)
> And that is how i think it should look like:
> ![image](https://user-images.githubusercontent.com/15233006/109414743-d0dab380-79b4-11eb-9884-8775fc47f1e3.png)

Ah, I see. I can manually wrap it to 80 characters real quick if that's all you need, just give me a minute.

Conni2461 commented on lua/telescope/pickers/layout_strategies.lua:
```diff
@@ -33,6 +34,24 @@
+---   - Flip the view of the current layout:
+---     - If using horizontal: if `true`, swaps the location of the
+---       results/prompt window and preview window
+---     - If using vertical: if `true, swaps the location of the results and
```
Missing `

smithbm2316 replied:
*sigh* i am tired and it is showing

Conni2461 replied:
Doesn't matter. I am super happy right now. Seeing that docgen ci work. Last thing i found after that i merge

smithbm2316 replied:
:heart: I appreciate it haha, I'm so glad to see the docgen working well too!

Conni2461 commented:
Thank you very much for everything :D i like it :)

smithbm2316 commented:
> Thank you very much for everything :D i like it :)

Thank you ALL for being so responsive and kind! I'm so happy I've been able to help contribute to such a great project :)
